### PR TITLE
Add Mago as task

### PR DESCRIPTION
**Pull request by johnatas-x**
| Q             | A
| ------------- | ---
| Branch        | mago-task
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Documented?   | yes
| Fixed tickets | #1193

This PR adds a Mago task

Mago is a recent tool that includes a formatter (like phpcs), a linter (like phpmd), and a static analyzer (like phpstan).
It is known for its performance and is starting to be increasingly adopted (for example, in Drupal).
Its integration into GrumPHP seems valuable.

mago-grumphp

# New Task Checklist:

- [x] Are the dependencies added to the composer.json suggestions?
- [x] Is the doc/tasks.md file updated?
- [x] Are the task parameters documented?
- [x] Is the task registered in the tasks.yml file?
- [x] Does the task contains phpunit tests?
- [x] Is the configuration having logical allowed types?
- [x] Does the task run in the correct context?
- [x] Is the `run()` method readable?
- [x] Is the `run()` method using the configuration correctly?
- [x] Are all CI services returning green?

### CODE REVIEW

File: src/Task/Mago.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+ */
+class Mago extends AbstractExternalTask
+{
+
+    public static function getConfigurableOptions(): ConfigOptionsResolver
+    {
+        $resolver = new OptionsResolver();
+        $resolver->setDefaults([
+            'formatter' => true,
+            'formatter_options' => ['--staged'],
```
Comment (veewee) [inline src/Task/Mago.php]: staged should only be a default during pre-commit I suppose?

File: src/Task/Mago.php [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            'linter' => true,
+            'linter_options' => ['--staged'],
+            'analyzer' => true,
+            'analyzer_options' => ['--staged'],
+            'guard' => false,
+            'guard_options' => [],
+        ]);
+
+        $resolver->addAllowedTypes('formatter', ['bool']);
+        $resolver->addAllowedTypes('formatter_options', ['array']);
```
Comment (veewee) [inline src/Task/Mago.php]: In other tasks, we provide configuration keys that map the options of the CLI.
Doing it like this introduces a new way of configuring tasks, which I'dd like to avoid at this point.

File: src/Task/Mago.php [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function run(ContextInterface $context): TaskResultInterface
+    {
+        $config = $this->getConfig()->getOptions();
+
+        if ($config['formatter'] === false
```
Comment (veewee) [inline src/Task/Mago.php]: I was thinking: wouldn't it make more sense to introduce a task per type of command?
Something like following tasks:

- `mago_fmt`
- `mago_lint`
- `mago_analyze`
- `mago_guard`

Every task will have it's own set of configurable options.

**Review by veewee — COMMENTED** [review]
Thanks for the PR.
I think this task in current state is not in line with other tasks. I've added some remarks inline in the code. Feel free to stat a discussion.

**Comment by johnatas-x** [comment]
Hey @veewee,

Thanks for taking the time to review the PR and share your feedback.
Your comments align with the questions I raised in the [related issue](https://github.com/phpro/grumphp/issues/1193#issuecomment-4353677159).

To be fully transparent, after opening the PR I had already started reworking it to introduce 4 commands. I’ll update the PR this week with those changes; it should end up closer to how the SecurityChecker tasks are structured, which should resolve the three threads.

If you have any additional feedback, feel free to share it so I can address everything at once.

**Comment by johnatas-x** [comment]
Hey,

I just pushed a refactor splitting the `mago` task into 4 distinct subtasks: `mago_lint`, `mago_analyze`, `mago_format`, and `mago_guard`. Each task now follows the same structure as other tasks, with its own configurable options, dedicated documentation, and unit tests.

The `mago` task itself now serves two purposes: it acts as a redirect, pointing users to the 4 subtasks when invoked directly, and as a base class holding the shared logic to avoid code duplication across the subtasks.

Happy to hear any feedback — whether on the option coverage, naming, documentation, or anything else that could be improved.

File: src/Task/Mago.php [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+use GrumPHP\Task\Config\ConfigOptionsResolver;
+use GrumPHP\Task\Context\ContextInterface;
+use GrumPHP\Task\Context\GitPreCommitContext;
+use GrumPHP\Task\Context\RunContext;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+/**
+ * @extends AbstractExternalTask<ProcessFormatterInterface>
+ */
+class Mago extends AbstractExternalTask
```
Comment (veewee) [inline src/Task/Mago.php]: Maybe better to have it abstract as wel? There is no such thing as a mago task

File: src/Task/Mago.php [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+{
+
+    public static function getConfigurableOptions(): ConfigOptionsResolver
```
Comment (veewee) [inline src/Task/Mago.php]: not needed when abstract.

File: src/Task/Mago.php [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    {
+        return ConfigOptionsResolver::fromOptionsResolver(new OptionsResolver());
+    }
+
+    public function canRunInContext(ContextInterface $context): bool
+    {
+        return $context instanceof GitPreCommitContext || $context instanceof RunContext;
+    }
+
+    public function run(ContextInterface $context): TaskResultInterface
```
Comment (veewee) [inline src/Task/Mago.php]: not needed when abstract.

File: src/Task/Mago.php [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+            . '(https://github.com/phpro/grumphp/blob/master/doc/tasks/mago/linter.md)'.PHP_EOL
+        );
+    }
+
+    protected static function configureSharedOptions(OptionsResolver $resolver): void
+    {
+        $resolver->setDefaults([
+            'retain-codes' => [],
+            'ignore-baseline' => null,
+            'fix' => null,
```
Comment (veewee) [inline src/Task/Mago.php]: lets not add the fix option.
Rather use something like this:

https://github.com/phpro/grumphp/blob/732a228acacd062faec211e0e3e4317ce4ea9bd6/src/Task/PhpCsFixer.php#L107-L116

By default grumphp does dry-run.
When failed, it asks if you want it to be fixed or not (meaning it 'll run the fix command)

File: src/Task/Mago.php [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+            'fail-on-remaining' => null,
+            'sort' => null,
+            'fixable-only' => null,
+            'reporting-format' => null,
+            'reporting-target' => null,
+            'minimum-report-level' => null,
+            'minimum-fail-level' => null,
+            'dry-run' => null,
```
Comment (veewee) [inline src/Task/Mago.php]: remove dry-run option.

File: src/Task/Mago.php [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+        $resolver->addAllowedValues('reporting-format', [
+            null, 'rich', 'medium', 'short', 'ariadne', 'github', 'gitlab',
+            'json', 'count', 'code-count', 'checkstyle', 'emacs', 'sarif',
+        ]);
+        $resolver->addAllowedValues('reporting-target', [null, 'stdout', 'stderr']);
+        $resolver->addAllowedValues('minimum-report-level', [null, 'note', 'help', 'warning', 'error']);
+        $resolver->addAllowedValues('minimum-fail-level', [null, 'note', 'help', 'warning', 'error']);
+    }
+
+    protected function resolveFixOption(array $config): ?string
```
Comment (veewee) [inline src/Task/Mago.php]: not needed. See comment above

File: src/Task/Mago.php [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    {
+        return $config['fix'];
+    }
+
+    protected function validateFixCompatibility(
```
Comment (veewee) [inline src/Task/Mago.php]: I dont think we'll need this given that we'll have control about building the fix command from within the grumphp tasks directly with the approach above.

File: src/Task/Mago.php [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+        $arguments->add('--fix');
+        $arguments->addOptionalArgument('--potentially-unsafe', 'potentially-unsafe' === $fix);
+        $arguments->addOptionalArgument('--unsafe', 'unsafe' === $fix);
+        $arguments->addOptionalArgument('--fail-on-remaining', $config['fail-on-remaining']);
+        $arguments->addOptionalArgument('--dry-run', $config['dry-run']);
+    }
+
+    protected function addSharedArguments(
+        ProcessArgumentsCollection $arguments,
+        array $config
```
Comment (veewee) [inline src/Task/Mago.php]: add a psalm-type (shape) maybe to config? (same goes for `addFixArguments`)

File: src/Task/Mago.php [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+            default => null,
+        };
+
+        return null !== $error ? TaskResult::createFailed($this, $context, $error) : null;
+    }
+
+    protected function addFixArguments(
+        ProcessArgumentsCollection $arguments,
+        array $config,
+        ?string $fix
```
Comment (veewee) [inline src/Task/Mago.php]: no need for the fix var.

File: src/Task/MagoAnalyzer.php [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+class MagoAnalyzer extends Mago
+{
+
+    public static function getConfigurableOptions(): ConfigOptionsResolver
+    {
+        $resolver = new OptionsResolver();
+        $resolver->setDefaults([
+            'no-stubs' => null,
+            'staged' => null,
```
Comment (veewee) [inline src/Task/MagoAnalyzer.php]: does it make sense to have `staged` here?
We could possibly automatically do this during pre-commit and dont do it during run context?

That way we don't have to pass files ever.. Otherwise we'dd need to pass files during pre-commit.

File: src/Task/MagoFormatter.php [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+
+    public static function getConfigurableOptions(): ConfigOptionsResolver
+    {
+        $resolver = new OptionsResolver();
+        $resolver->setDefaults([
+            'type' => 'default',
+        ]);
+
+        $resolver->addAllowedTypes('type', ['string']);
+        $resolver->addAllowedValues('type', ['default', 'dry-run', 'check', 'staged']);
```
Comment (veewee) [inline src/Task/MagoFormatter.php]: lets not hide options behind a custom 'type'

File: src/Task/MagoGuard.php [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        $resolver = new OptionsResolver();
+        $resolver->setDefaults([
+            'no-stubs' => null,
+            'checks' => 'all',
+        ]);
+
+        $resolver->addAllowedTypes('no-stubs', ['null', 'bool']);
+        $resolver->addAllowedTypes('checks', ['string']);
+        $resolver->addAllowedValues('checks', ['all', 'structural', 'perimeter']);
```
Comment (veewee) [inline src/Task/MagoGuard.php]: lets not hide options behind a custom 'checks'. make them bools.

File: src/Task/MagoLinter.php [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+{
+
+    public static function getConfigurableOptions(): ConfigOptionsResolver
+    {
+        $resolver = new OptionsResolver();
+        $resolver->setDefaults([
+            'semantics' => null,
+            'pedantic' => null,
+            'only' => [],
+            'staged' => null,
```
Comment (veewee) [inline src/Task/MagoLinter.php]: see previous feedback: lets make staged or not dependant of the context.

File: src/Task/Mago.php [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+            . '- mago_format '
+            . '(https://github.com/phpro/grumphp/blob/master/doc/tasks/mago/formatter.md)'.PHP_EOL
+            . '- mago_guard '
+            . '(https://github.com/phpro/grumphp/blob/master/doc/tasks/mago/guard.md)'.PHP_EOL
+            . '- mago_lint '
+            . '(https://github.com/phpro/grumphp/blob/master/doc/tasks/mago/linter.md)'.PHP_EOL
+        );
+    }
+
+    protected static function configureSharedOptions(OptionsResolver $resolver): void
```
Comment (veewee) [inline src/Task/Mago.php]: are those options actually shared across all the commands in mago as well?
Because this could bite us if not. For example: see formatter tasks, which is not using the shared options.

**Comment by veewee** [comment]
Thanks for the changes.
I currently only looked at the tasks and added some comments.
In general, I think we should do things a bit different here and there.

**Comment by johnatas-x** [comment]
No worries
Did you add comments? I can’t see any

**Comment by johnatas-x** [comment]
Hi @veewee,

Thanks for all your feedback!

I'm working through it bit by bit. I'll push a new version this weekend that incorporates your comments, and I'll also include answers to your questions.

Talk to you soon.

**Comment by johnatas-x** [comment]
Hi, here is a summary of all the changes made based on your feedback.

---

### `staged` option removed from config — auto-detected from context

The `staged` flag is no longer a user-facing option. It is now automatically passed to `mago lint` / `mago analyze` when running in a `GitPreCommitContext`, and omitted in `RunContext`. This matches GrumPHP's philosophy of letting the context drive behavior rather than the user config.

On that note: I kept `--staged` (mago's native flag) rather than passing file paths from the diff context. The reason is that files and directories processed by Mago are defined in `mago.toml`. Passing explicit paths from GrumPHP's diff would *override* that configuration — meaning a file not covered by `mago.toml` could end up being linted/analyzed/formatted unintentionally. Using `--staged` delegates the file selection entirely to Mago's own configuration, which is the safe and expected behavior.

---

### `mago lint` / `mago guard` — always run in dry-run + fix mode

These two tasks always run with `--fix --dry-run --fail-on-remaining`. When the task fails, GrumPHP offers to re-run with `--fix` applied (without `--dry-run`). The fix aggressiveness is controlled by a new `fix-mode` option (`safe` by default, `potentially-unsafe`, `unsafe`).

---

### `mago analyze` — runs directly for full diagnostic output

The analyzer runs `mago analyze` without `--fix --dry-run`, as combining those flags suppresses the detailed diagnostic output. When the task fails, GrumPHP offers a fix command (`mago analyze --fix`) built separately, with the appropriate fix-mode flag if configured.

---

### Removed options incompatible with `--fix`

`fixable-only`, `reporting-format`, `reporting-target`, and `minimum-fail-level` have been removed from the shared options as they cannot be used alongside `--fix`. `fail-on-remaining` is now hardcoded (always on) since it is required to detect unfixed issues during the dry-run step.

---

### `mago_formatter` — always runs `--dry-run`, offers fix in both contexts

The `type` option has been removed. The formatter always runs with `--dry-run`. In both `GitPreCommitContext` and `RunContext`, GrumPHP offers to re-run without `--dry-run` to apply formatting in-place.

---

### `mago_guard` — `checks` replaced by `structural` and `perimeter` bools

Following your feedback, the custom `checks` string option has been replaced by two independent boolean flags `structural` and `perimeter`, consistent with how other options are handled across Mago tasks.

---

### Abstract `Mago` class kept

I kept the `Mago` abstract class, which is extended by `MagoAnalyzer`, `MagoLinter`, and `MagoGuard`. These three share the same execution pattern (shared options, `--fix --dry-run`, `createFailedWithFix`). `MagoFormatter` now extends `AbstractExternalTask` directly as it has its own distinct behavior and does not benefit from the shared abstract.

> Are you happy with this approach, or would you prefer I remove the abstract and duplicate its content across the three tasks?

---

I ran a few tests on a real project and things seem to work as expected. That said, my Mago configuration is fairly minimal, so edge cases with more advanced setups (custom rules, baseline files, analyzer stubs, etc.) haven't been covered. Additional testing from someone with a more complete Mago configuration would be welcome.

Happy to address any further feedback, thanks!

**Comment by veewee** [comment]
Hey @johnatas-x,

Rather than going back and forth on the open threads another dozen times, I pushed a version of how I think the Mago tasks could look. Easier to react to something concrete than to keep describing it.

The biggest change: I dropped the abstract `Mago` class. It was trying to share logic across all four commands, but the commands aren't really that alike. Guard can't auto-fix, format handles staged files differently, and the linter/analyzer don't even run the same way. So the base class ended up carrying behaviour that only fit some of them, which made the whole thing harder to follow than it should be. Each command is now a standalone task you can read top to bottom.

While I was in there I lined the tasks up with how the rest of GrumPHP behaves (dry-run by default, offer to fix when it makes sense) and with what Mago actually supports. I went through every flag against the real `mago` CLI, so guard no longer pretends it can fix things, and the structural/perimeter choice is now a single `mode` option since you can't use both at once anyway.

Tests cover each option per task and the fix commands they generate, and the docs are updated to match.

Could you pull it and try it on a real project, especially the pre-commit runs? Want to make sure it behaves the way you'd expect on a setup more complete than mine. If it feels right to you, I'll get it merged.

**Comment by johnatas-x** [comment]
Hi @veewee,

Thank you for the effort you've put into this.

I'm going to review your changes to better understand the approach and improve my own work going forward.

I'll test the branch on three real projects during the course of next week and will get back to you with feedback based on those tests.

**Comment by veewee** [comment]
Tested it out locally and seems to do its thing :)

**Comment by johnatas-x** [comment]
Hello,

I can confirm that it's working (tested on three real-world projects).

My tests were carried out with very simple configurations, which I'll gradually harden over time. If I encounter any issues or identify opportunities for improvement, I'll be happy to provide feedback and submit enhancements as I go.

I think we can close the related issue.

Thanks a lot for your work!
